File: database/src/database-test/dreammaker/android/expensetracker/database/dao/AccountDaoTest.java
package dreammaker.android.expensetracker.database.dao;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Arrays;
import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import dreammaker.android.expensetracker.database.ExpensesDatabase;
import dreammaker.android.expensetracker.database.entity.Account;
import dreammaker.android.expensetracker.database.model.AccountModel;
import dreammaker.android.expensetracker.database.type.Currency;
import dreammaker.android.expensetracker.database.util.DatabaseUtil;

import static dreammaker.android.expensetracker.database.util.DatabaseUtil.createAccount;
import static dreammaker.android.expensetracker.database.util.DatabaseUtil.createAccountModel;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class AccountDaoTest {

    ExpensesDatabase db;

    AccountDao dao;

    @Before
    public void openDb(){
        db = DatabaseUtil.createInMemoryExpenseDatabase(7);
        dao = db.getAccountDao();
    }

    @After
    public void closeDb() {
        DatabaseUtil.closeDBSilently(db);
    }

    @Test
    public void addAccount() {
        Account account = createAccount(0L,"New Account", Currency.ZERO);
        long id = dao.addAccount(account);
        assertTrue(id>0);
    }

    @Test
    public void updateAccount() {
        Account account = createAccount(1L,"acc 1 updated",Currency.ZERO);
        int changes = dao.updateAccount(account);
        assertEquals(1,changes);
    }

    @Test
    public void getAllAccountsLive() throws Exception {
        LiveData<List<AccountModel>> liveData = dao.getAllAccountsLive();
        List<AccountModel> actual = DatabaseUtil.getValueFromLivedata(liveData,5000);

        AccountModel account1 = createAccountModel(1L,"acc 1",Currency.valueOf("-100"));
        AccountModel account2 = createAccountModel(2L,"acc 2",Currency.valueOf("3500"));
        AccountModel account3 = createAccountModel(3L,"acc 3",Currency.valueOf("25000"));
        List<AccountModel> expected = Arrays.asList(account1,account2,account3);

        assertEquals(expected,actual);
    }

    @Test
    public void getAllAccountsWithUsageCountLive() throws Exception {
        LiveData<List<AccountModel>> liveData = dao.getAllAccountWithUsageCountLive();
        List<AccountModel> actual = DatabaseUtil.getValueFromLivedata(liveData,5000);

        AccountModel account1 = createAccountModel(1L,"acc 1",Currency.valueOf("-100"),8);
        AccountModel account2 = createAccountModel(2L,"acc 2",Currency.valueOf("3500"),3);
        AccountModel account3 = createAccountModel(3L,"acc 3",Currency.valueOf("25000"),5);
        List<AccountModel> expected = Arrays.asList(account1,account2,account3);

        assertEquals(expected,actual);
    }

    @Test
    public void findAccountById() {
        Account expected = createAccount(1L,"acc 1",Currency.valueOf("-100"));
        Account actual = dao.findAccountById(1);
        assertEquals(expected,actual);
    }

    @Test
    public void removeAccounts() {
        int changes = dao.removeAccounts(new long[]{1,3});
        assertEquals(2,changes);
    }
}

File: database/src/androidTest/java/dreammaker/android/expensetracker/database/dao/AccountDaoTest.kt
package dreammaker.android.expensetracker.database.dao

import androidx.test.ext.junit.runners.AndroidJUnit4
import dreammaker.android.expensetracker.database.ExpensesDatabase
import dreammaker.android.expensetracker.database.createInMemoryDB
import dreammaker.android.expensetracker.database.fake.FakeDataCallbacks
import dreammaker.android.expensetracker.database.model.AccountEntity
import junit.framework.TestCase.assertEquals
import junit.framework.TestCase.assertTrue
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.runTest
import org.junit.After
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.time.LocalDateTime

@RunWith(AndroidJUnit4::class)
class AccountDaoTest {

    lateinit var db: ExpensesDatabase
    lateinit var dao: AccountDao


    @Before
    fun setUp() {
        db = createInMemoryDB(FakeDataCallbacks.getCallbackForCurrentDBVersion())
        dao = db.accountDao
    }

    @After
    fun tearDown() {
        db.close()
    }

    @Test
    fun getFlowAccountById() = runTest {
        val actual = dao.findAccountByIdFlow(1).first()
        val expected = AccountEntity(1,"Account 1",150.0, LocalDateTime.of(2025,10,15,15,56,20),1)
        assertEquals(expected,actual)
    }

    @Test
    fun insert() = runTest {
        val account = AccountEntity(
            0,
            "Test Account Coroutine",
            200.0,
            LocalDateTime.of(2026,2,1,0,0,0,0),
            2)
        val id = dao.insert(account)

        val actual = dao.findAccountByIdFlow(id).first()
        val expected = account.copy(id=id)

        assertTrue("id must be positive number", id > 0)
        assertEquals("inserted account not found by id", expected, actual)
    }

    @Test
    fun insertMultiple() = runTest {
        val accounts = listOf(
            AccountEntity(0, "Multi 1", 100.0, null, null),
            AccountEntity(0, "Multi 2", 200.0, null, null)
        )
        val ids = dao.insertMultiple(accounts)
        assertEquals(2, ids.size)
        
        val all = dao.getAccountsFlow().first()
        // Account 1, Account 2 are already there from FakeData
        assertTrue(all.any { it.name == "Multi 1" })
        assertTrue(all.any { it.name == "Multi 2" })
    }

    @Test
    fun getAccountsFlow() = runTest {
        val actual = dao.getAccountsFlow().first()
        // Fake data has "Account 1" and "Account 2"
        assertEquals(2, actual.size)
        assertEquals("Account 1", actual[0].name)
        assertEquals("Account 2", actual[1].name)
    }

    @Test
    fun update() = runTest {
        val account = AccountEntity(1,"Account 1 Updated",160.0, LocalDateTime.now(),2)
        dao.update(account)
        val actual = dao.findAccountByIdFlow(1).first()
        assertEquals("Account 1 Updated", actual?.name)
    }

    @Test
    fun deleteById() = runTest {
        val changes = dao.deleteById(1)
        assertEquals(1, changes)
        val account = dao.findAccountByIdFlow(1).first()
        assertNull(account)
    }

    @Test
    fun deleteMultipleByIds() = runTest {
        val changes = dao.deleteMultipleByIds(listOf(1,2))
        assertEquals(2, changes)
        val all = dao.getAccountsFlow().first()
        assertTrue(all.isEmpty())
    }
}
